bin/plugin/commands/readme.js

```diff
@@ -52,112 +38,136 @@ exports.options = [
  */
 exports.handler = async ( opt ) => {
 	await updateReadme( {
-		owner: config.githubRepositoryOwner,
-		repo: config.githubRepositoryName,
-		milestone: opt.milestone,
-		path: opt.path,
+		plugin: opt.plugin,
 		token: opt.token,
 	} );
 };
 
 /**
- * Detects the path to the readme.txt file to update based on the milestone title.
- *
- * @param {string} milestone Milestone title.
+ * Updates the `readme.txt` file with the given changelog.
  *
- * @return {string} Detected readme.txt path.
+ * @param {string} readmeFile Readme file path.
+ * @param {string} changelog  Changelog in markdown.
+ * @return {string} Success status message.
  */
-function detectReadmePath( milestone ) {
-	const slug = milestone.match( /^([a-z0-9-]+) / );
-	if ( ! slug ) {
-		throw new Error(
-			`The ${ milestone } milestone does not start with a valid plugin slug.`
-		);
+function updateReadmeChangelog( readmeFile, changelog ) {
+	const fileContent = fs.readFileSync( readmeFile, 'utf-8' );
+
+	const stableTagVersionMatches = fileContent.match(
+		/^Stable tag:\s*(\d+\.\d+\.\d+)$/m
+	);
+	if ( ! stableTagVersionMatches ) {
+		throw new Error( `Unable to locate stable tag in ${ readmeFile }` );
 	}
+	const stableTagVersion = stableTagVersionMatches[ 1 ];
+
+	const regex = new RegExp(
+		`(== Changelog ==\n+)(= ${ stableTagVersion } =\n+)([^=]+)`
+	);
+
+	const versionHeading = `= ${ stableTagVersion } =\n\n`;
+	const normalizedChangelog = changelog.trimEnd() + '\n';
+
+	let status = '';
+
+	// Try to merge the new changelog with the existing changelog.
+	let newContent = fileContent.replace(
+		regex,
+		( match, changelogHeading, _versionHeading, existingChangelog ) => {
+			status =
+				'Merged existing changelog with the new changelog in an Other section.';
+			return `${ changelogHeading }${ _versionHeading }${ normalizedChangelog }\n**Other**\n\n${ existingChangelog }`;
+		}
+	);
```

Comment:
Does this assume that any existing changelog entries should be classified as "Other"? If you have an existing `**Enhancements**` section in the changelog and then run this command again, you would end up with something like this:

```md
**Enhancements**

* Avoid needless array allocation in rgb to hex conversion. ([1104](https://github.com/WordPress/performance/pull/1104))

**Bug Fixes**

* Fix Imagick detecting partial transparency. ([1215](https://github.com/WordPress/performance/pull/1215))

**Other**

**Enhancements**

* Improve overall code quality with stricter static analysis checks. ([775](https://github.com/WordPress/performance/issues/775))
```

This is probably fine, so long as people are paying attention to the output and not merging it uncritically, just something I noticed. I also noticed that since there were no _additional_ PRs to add to the changelog, the auto-sizes changelog didn't get the `**Other**` heading added above the existing `Improve overall code quality with stricter static analysis checks.`. Also, probably fine, but wonder if we should try to be consistent—even if there are no other merged PRs to add to the changelong.

Reply:
Yeah, it will still require the user to review the output. But the important thing here is that it doesn't blow away the existing changelog entries already in the readme, so at least it is easier for the developer to review the changes and merge them.

Reply:
@westonruter Chiming in late here, but I'm not sure I understand the purpose of this. What would "Other" be for? Shouldn't we be good with features, enhancements, and bug fixes?

Reply:
@felixarntz It's whatever was originally added to that section in the readme. It's not intended to be left as-is, so when reviewing the changes the other items can be moved into the other sections.